components/brave_vpn/browser/connection/win/brave_vpn_helper/vpn_dns_handler.cc

```diff
@@ -0,0 +1,194 @@
+    return true;
+  }
+  bool success = CloseEngineSession();
+  if (success) {
+    engine_ = nullptr;
+  }
+  return success;
+}
+
+internal::CheckConnectionResult VpnDnsHandler::GetVpnEntryStatus() {
+  VLOG(1) << __func__;
+  if (connection_result_for_testing_.has_value()) {
+    return connection_result_for_testing_.value();
+  }
+  return internal::CheckConnection(
+      base::UTF8ToWide(brave_vpn::kBraveVPNEntryName));
+}
+
+void VpnDnsHandler::UpdateFiltersState() {
+  VLOG(1) << __func__;
+  switch (GetVpnEntryStatus()) {
+    case internal::CheckConnectionResult::CONNECTED:
+      VLOG(1) << "BraveVPN connected, set filters";
+      if (IsActive()) {
+        VLOG(1) << "Filters are already installed";
+        return;
+      }
+      if (!SetFilters(base::UTF8ToWide(brave_vpn::kBraveVPNEntryName))) {
+        VLOG(1) << "Failed to set DNS filters";
+        auto result = internal::DisconnectEntry(
+            base::UTF8ToWide(brave_vpn::kBraveVPNEntryName));
+        if (!result.success) {
+          VLOG(1) << "Failed to disconnect entry:" << result.error_description;
+        }
+        Exit();
+        return;
+      }
+      break;
+    case internal::CheckConnectionResult::DISCONNECTED:
+      VLOG(1) << "BraveVPN Disconnected, remove filters";
+      if (!RemoveFilters(base::UTF8ToWide(brave_vpn::kBraveVPNEntryName))) {
+        VLOG(1) << "Failed to remove DNS filters";
+        Exit();
+        break;
+      }
+      // Reset service launch counter if dns filters successfully removed.
+      ResetLaunchCounter();
+      Exit();
+      break;
+    default:
+      VLOG(1) << "BraveVPN is connecting, try later after "
+              << kCheckConnectionIntervalInSeconds << " seconds.";
+      break;
+  }
+}
+
+void VpnDnsHandler::CloseWatchers() {
+  if (event_handle_for_vpn_) {
+    CloseHandle(event_handle_for_vpn_);
+    event_handle_for_vpn_ = nullptr;
+  }
+  periodic_timer_.Stop();
+}
+
+void VpnDnsHandler::Exit() {
+  CloseWatchers();
+  delegate_->SignalExit();
+}
+
+void VpnDnsHandler::OnObjectSignaled(HANDLE object) {
+  VLOG(1) << __func__;
+  // We receive events from all connections in the system and filter here
+  // only expected brave vpn event.
+  if (object != event_handle_for_vpn_)
+    return;
+  UpdateFiltersState();
+}
+
+void VpnDnsHandler::SubscribeForRasNotifications(HANDLE event_handle) {
+  VLOG(1) << __func__;
+  if (!SubscribeRasConnectionNotification(event_handle)) {
+    VLOG(1) << "Failed to subscripbe for vpn notifications";
+  }
+}
+
+void VpnDnsHandler::StartVPNConnectionChangeMonitoring() {
+  DCHECK(!event_handle_for_vpn_);
+  DCHECK(!IsActive());
+
+  event_handle_for_vpn_ = CreateEvent(NULL, false, false, NULL);
+  SubscribeForRasNotifications(event_handle_for_vpn_);
+
+  connected_disconnected_event_watcher_.StartWatchingMultipleTimes(
+      event_handle_for_vpn_, this);
+
+  periodic_timer_.Start(FROM_HERE,
+                        base::Seconds(kCheckConnectionIntervalInSeconds),
```

Comment:
there is already a subscription to RasNotification from the endpoint, the polling is used as an additional protection step and looks more reliable because if the vpn connection (or driver goes down due to crash) and we miss the notification, the user won't have internet.

Reply:
I thought using `NetworkChangeNotifier::NetworkChangeObserver` is sufficient for getting notification whenever vpn connection changes becuase vpn connection change causes os network status change.
Of course, 3s interval polling doesn't have performance impact. So I think current is also fine. but I just wanted to mention that `NetworkChangeNotifier::NetworkChangeObserver` also works nicely at `BraveVPNOSConnectionAPIBase` for same purpose.